src/flows/pipes/Notification/view.tsx

```diff
@@ -0,0 +1,415 @@
+  }
+
+  const hasTaskOption = useMemo(
+    () =>
+      !!Object.keys(
+        remove(
+          parse(simplify(queryText)),
+          node =>
+            node.type === 'OptionStatement' &&
+            node.assignment.id.name === 'task'
+        ).reduce((acc, curr) => {
+          // eslint-disable-next-line no-extra-semi
+          ;(curr.assignment?.init?.properties || []).reduce((_acc, _curr) => {
+            if (_curr.key?.name && _curr.value?.location?.source) {
+              _acc[_curr.key.name] = _curr.value.location.source
+            }
+
+            return _acc
+          }, acc)
+
+          return acc
+        }, {})
+      ).length,
+    [queryText]
+  )
+
+  const numericColumns = (results.parsed.table?.columnKeys || []).filter(
+    key => {
+      if (key === 'result' || key === 'table') {
+        return false
+      }
+
+      const columnType = results.parsed.table.getColumnType(key)
+
+      return columnType === 'time' || columnType === 'number'
+    }
+  )
+
+  const updateMessage = evt => {
+    update({
+      message: evt.target.value,
+    })
+  }
+
+  const updateInterval = evt => {
+    update({
+      interval: evt.target.value,
+    })
+  }
+
+  const updateOffset = evt => {
+    update({
+      offset: evt.target.value,
+    })
+  }
+
+  const updateEndpoint = which => {
+    update({
+      endpoint: which,
+      endpointData: DEFAULT_ENDPOINTS[which].data,
+    })
+  }
+
+  const warningMessage = useMemo(() => {
+    if (!hasTaskOption) {
+      return
+    }
+
+    return (
+      <div className="flow-error">
+        <div className="flow-error--header">
+          <Icon
+            glyph={IconFont.AlertTriangle}
+            className="flow-error--vis-toggle"
+          />
+        </div>
+        <div className="flow-error--body">
+          <h1>The task option is reserved</h1>
+          <p>
+            This panel will take precedence over any task configuration and
+            overwrite the option. Remove it from your source query to remove
+            this message
+          </p>
+        </div>
+      </div>
+    )
+  }, [hasTaskOption])
+
+  const avail = Object.keys(DEFAULT_ENDPOINTS).map(k => (
+    <Tabs.Tab
+      key={k}
+      id={k}
+      onClick={() => updateEndpoint(k)}
+      text={DEFAULT_ENDPOINTS[k].name}
+      active={data.endpoint === k}
+    />
+  ))
+
+  const generateTask = useCallback(() => {
+    // simplify takes care of all the variable nonsense in the query
+    const ast = parse(simplify(queryText))
+
+    const vars = remove(
+      ast,
+      node => node.type === 'OptionStatement' && node.assignment.id.name === 'v'
+    ).reduce((acc, curr) => {
+      // eslint-disable-next-line no-extra-semi
+      ;(curr.assignment?.init?.properties || []).reduce((_acc, _curr) => {
+        if (_curr.key?.name && _curr.value?.location?.source) {
+          _acc[_curr.key.name] = _curr.value.location.source
+        }
+
+        return _acc
+      }, acc)
```

Comment:
This is used twice in this specific view, and is likely used in other parts of notebooks. Should we abstract this into a utility function to reduce potential drift?

Reply:
i've typed this a few times and end up going on a long rant about misconceptions about reusability / DRY methodologies when they're cargo culted. imma try to keep it on point, lemme know if i missed the mark as it spreads kind of generally.

even if two pieces of code are exactly the same, character for character, they might not have the same intent behind them. By bundling the code together, you also bundle the intent. The issue arises when complexity grows in the app, and these intents diverge. You start seeing edge cases and conditional treatment of the logic within the function (if x do y, or else do z), and then you start maintaining those edge cases that you introduced in the name of making the code easier. DRY for DRY's sake turns into reusing nonreusable code, or front loading the complexity of managing that complexity. even in this example, one just wants to know if a task option is there, the other wants to remove all task options from the source query and convert them to a POJO for later manipulation. These different intents can have vastly different implementations, but by joining them into a util function, we now have to settle on the implementation that satisfies both use cases.

DRY has it's merits, and there's important concepts to learn from it, but don't buy the astroturf. copypasta gang for life.